jailer.go

```diff
@@ -0,0 +1,451 @@
+
+// secure computing levels
+const (
+	// SeccompLevelDisable is the default value.
+	SeccompLevelDisable SeccompLevelValue = iota
+	// SeccompLevelBasic prohibits syscalls not whitelisted by Firecracker.
+	SeccompLevelBasic
+	// SeccompLevelAdvanced adds further checks on some of the parameters of the
+	// allowed syscalls.
+	SeccompLevelAdvanced
+)
+
+// JailerConfig is jailer specific configuration needed to execute the jailer.
+type JailerConfig struct {
+	// GID the jailer switches to as it execs the target binary.
+	GID *int
+
+	// UID the jailer switches to as it execs the target binary.
+	UID *int
+
+	// ID is the unique VM identification string, which may contain alphanumeric
+	// characters and hyphens. The maximum id length is currently 64 characters
+	ID string
+
+	// NumaNode represents the NUMA node the process gets assigned to.
+	NumaNode *int
+
+	// ExecFile is the path to the Firecracker binary that will be exec-ed by
+	// the jailer. The user can provide a path to any binary, but the interaction
+	// with the jailer is mostly Firecracker specific.
+	ExecFile string
+
+	// ChrootBaseDir represents the base folder where chroot jails are built. The
+	// default is /srv/jailer
+	ChrootBaseDir string
+
+	// NetNS represents the path to a network namespace handle. If present, the
+	// jailer will use this to join the associated network namespace
+	NetNS string
+
+	//  Daemonize is set to true, call setsid() and redirect STDIN, STDOUT, and
+	//  STDERR to /dev/null
+	Daemonize bool
+
+	// SeccompLevel specifies whether seccomp filters should be installed and how
+	// restrictive they should be. Possible values are:
+	//
+	//	0 : (default): disabled.
+	//	1 : basic filtering. This prohibits syscalls not whitelisted by Firecracker.
+	//	2 : advanced filtering. This adds further checks on some of the
+	//			parameters of the allowed syscalls.
+	SeccompLevel SeccompLevelValue
+
+	// DevMapperStrategy will dictate how files are transfered to the root drive.
+	DevMapperStrategy HandlersAdaptor
+}
+
+// JailerCommandBuilder will build a jailer command. This can be used to
+// specify that a jailed firecracker executable wants to be run on the Machine.
+type JailerCommandBuilder struct {
+	bin      string
+	id       string
+	uid      int
+	gid      int
+	execFile string
+	node     int
+
+	// optional params
+	chrootBaseDir string
+	netNS         string
+	daemonize     bool
+	seccompLevel  SeccompLevelValue
+
+	stdin  io.Reader
+	stdout io.Writer
+	stderr io.Writer
+}
+
+// Args returns the specified set of args to be used
+// in command construction.
+func (b JailerCommandBuilder) Args() []string {
+	args := []string{}
+	args = append(args, b.ID()...)
+	args = append(args, b.UID()...)
+	args = append(args, b.GID()...)
+	args = append(args, b.ExecFile()...)
+	args = append(args, b.NumaNode()...)
+
+	if len(b.chrootBaseDir) > 0 {
+		args = append(args, b.ChrootBaseDir()...)
+	}
+
+	if len(b.netNS) > 0 {
+		args = append(args, b.NetNS()...)
```

Comment:
```args = append(args, []string{"--net-ns", b.netNS}```
would be a lot more understandable than adding an extra func for each argument

Reply:
It allows for defaults if they are function based and hides that from the `Build` method allowing for a simpler view what what is being built. Maybe a better name rather than `NetNS`, maybe `GetCommandArgsNetNS` or something?

Reply:
I'm not sure I follow. You check defaults with `if len(b.netNS) > 0` and `NetNS` not called anywhere else except here.
So basically this can be
```go
struct Builder {
  NetNS string
}

func (b*builder) Build() string {
  if NetNS != "" {
    args = append(args, "--net-ns", b.netNS)
  }
}
```

Reply:
I can add a constructor that will solve the defaults issue. Currently there is no constructor, so users would always have to set some value.

Reply:
Yep. I think constructor is a good place to init defaults 👍